### PR TITLE
[Impeller] Round out extreme angles between curve polyline segments.

### PR DESCRIPTION
**Pull request by bdero**
Resolves https://github.com/flutter/flutter/issues/137956.

Detect when the angle between two curve segments is > 10 degrees and insert smoothing geometry if so. See [this comment](https://github.com/flutter/flutter/issues/137956#issuecomment-2148395911) for the full rationale behind this change. 

```dart
    Paint paint = Paint()
      ..color = Colors.white
      ..strokeWidth = 50
      ..strokeCap = StrokeCap.round
      ..strokeJoin = StrokeJoin.round
      ..style = PaintingStyle.stroke;

    Path path = Path();
    path.moveTo(0, 0);
    path.quadraticBezierTo(100, 100, 0, 0);
    canvas.drawPath(path, paint);
```

Before:
image

After:
image

### CODE REVIEW

**Comment by bdero** [comment]
Hmm, looks like this fix introduces a regression with `CanRenderStrokePathThatEndsAtSharpTurn`

image

**Comment by bdero** [comment]
5 degrees was way too sensitive a threshold. Bumped down to 10 degrees.

File: impeller/entity/geometry/stroke_path_geometry.cc [hunk 1]
```diff
@@ -155,19 +156,52 @@ class StrokeGenerator {
...
     }
   }
 
+  /// @brief  Represents a ray in 2D space.
+  ///
+  ///         This is a simple convenience struct for handling polyline offset
+  ///         values when generating stroke geometry. For performance reasons,
+  ///         it's sometimes adventageous to track the direction and magnitude
+  ///         for offsets separately.
+  struct Ray {
```
Comment (chinmaygarde) [inline impeller/entity/geometry/stroke_path_geometry.cc:172]: This could perhaps go in //impeller/geometry.

**Review by jonahwilliams — APPROVED** [review]
LGTM!

**Review by chinmaygarde — APPROVED** [review]

File: impeller/entity/geometry/stroke_path_geometry.cc (hunk 1 above)
Comment (bdero) [reply impeller/entity/geometry/stroke_path_geometry.cc]: Not opposed to throwing this in geometry. I'll do so in a follow-up. Also occurs to me that `Ray` is an inaccurate name for this... Should be something like `SeparatedVector2`.

**Pull request merged**